Dockerfile

```diff
@@ -1,16 +1,23 @@
-FROM node:21
-WORKDIR /usr/src/app
-
 ARG PORT=5000
 
+FROM node:21-alpine AS builder
+WORKDIR /usr/src/app
 RUN npm install -g pnpm
-
 COPY package*.json ./
-COPY pnpm-lock.yaml ./
+COPY tsconfig.json ./
 RUN pnpm install
-
-COPY . .
+COPY src ./src
+COPY scripts ./scripts
+COPY assets/images/characters ./assets/images/characters
 RUN npm run build
 
+FROM node:21-alpine AS main
+WORKDIR /usr/src/app
+RUN npm install -g pnpm
+COPY package*.json ./
+RUN pnpm install --production --quiet && rm -rf ~/.local/share && rm -rf ~/.cache
```

Comment:
I see you install pnpm and run pnpm install in both stages, that can increase build time significantly, so how about this:
  - Add a stage `base` at the top and install pnpm inside, then use it to create `builder` and `main` stages so that they inherit installed pnpm 
  - Copy `node_modules` from `builder` to `main` and run `pnpm prune --prod` instead of installing again